src/time_steppers.jl

```diff
@@ -45,237 +42,224 @@ time_step!(model; Nt, Δt, kwargs...) = time_step!(model, Nt, Δt; kwargs...)
-function calculate_GT!(GT, grid, closure, U, Φ, pHY′, K, F, parameters, time)
+""" Calculate the right-hand-side of the tracer advection-diffusion equation. """
+function calculate_Gc!(Gc, grid, closure, c, tracer_idx, U, C, K, Fc, parameters, time)
     @loop for k in (1:grid.Nz; (blockIdx().z - 1) * blockDim().z + threadIdx().z)
         @loop for j in (1:grid.Ny; (blockIdx().y - 1) * blockDim().y + threadIdx().y)
             @loop for i in (1:grid.Nx; (blockIdx().x - 1) * blockDim().x + threadIdx().x)
-                @inbounds GT[i, j, k] = (-div_flux(grid, U.u, U.v, U.w, Φ.T, i, j, k)
-                                            + ∇_κ_∇T(i, j, k, grid, Φ.T, closure, K)
-                                            + F.T(i, j, k, grid, time, U, Φ, parameters))
+                @inbounds Gc[i, j, k] = (-div_flux(grid, U.u, U.v, U.w, c, i, j, k)
+                                            + ∇_κ_∇c(i, j, k, grid, c, tracer_idx, closure, K)
+                                            + Fc(i, j, k, grid, time, U, C, parameters))
             end
         end
     end
+    return nothing
 end
 
-""" Calculate the right-hand-side of the salinity advection-diffusion equation. """
-function calculate_GS!(GS, grid, closure, U, Φ, pHY′, K, F, parameters, time)
-    @loop for k in (1:grid.Nz; (blockIdx().z - 1) * blockDim().z + threadIdx().z)
-        @loop for j in (1:grid.Ny; (blockIdx().y - 1) * blockDim().y + threadIdx().y)
-            @loop for i in (1:grid.Nx; (blockIdx().x - 1) * blockDim().x + threadIdx().x)
-                @inbounds GS[i, j, k] = (-div_flux(grid, U.u, U.v, U.w, Φ.S, i, j, k)
-                                            + ∇_κ_∇S(i, j, k, grid, Φ.S, closure, K)
-                                            + F.S(i, j, k, grid, time, U, Φ, parameters))
-            end
-        end
+""" Store previous value of the source term and calculate current source term. """
+function calculate_interior_source_terms!(G, arch, grid, coriolis, closure, U, C, pHY′, K, F, parameters, time)
+
+    Bx, By, Bz = floor(Int, grid.Nx/Tx), floor(Int, grid.Ny/Ty), grid.Nz  # Blocks in grid
+
+    @launch device(arch) threads=(Tx, Ty) blocks=(Bx, By, Bz) calculate_Gu!(G.u, grid, coriolis, closure, U, C, K, F, 
+                                                                            pHY′, parameters, time)
+
+    @launch device(arch) threads=(Tx, Ty) blocks=(Bx, By, Bz) calculate_Gv!(G.v, grid, coriolis, closure, U, C, K, F, 
+                                                                            pHY′, parameters, time)
+
+    @launch device(arch) threads=(Tx, Ty) blocks=(Bx, By, Bz) calculate_Gw!(G.w, grid, coriolis, closure, U, C, K, F, 
+                                                                            parameters, time)
+
+    for tracer_idx in 1:length(C)
+        @inbounds Gc = G[tracer_idx+3]
+        @inbounds Fc = F[tracer_idx+3]
+        @inbounds  c = C[tracer_idx]
```

Comment:
Stuff like this makes me a bit nervous as it assumes the first three elements of `G` and `F` are `u, v, w`. But I think we've discussed this at length and indexing named tuples with numbers is fine for now.

Reply:
This fact is currently enforced by the `ModelForcing` constructor and `Timestepper` constructor. We can add a line to the `Model` constructor that enforces the ordering of the names.